### ISSUE TITLE
Change Rename interface to use python {} format string notation and functionality instead of %() and regex

### ISSUE THREAD

**Issue by dafrose** [nipy/nipype#3269]
### Summary

The current implementation of `nipype.interfaces.utility.base.Rename` uses the older python notation for formatting strings with `%()` and parses regex to find the individual fields. With the newer `{}` format string notation, the built-in `string.Formatter` class can be used to extract all fields in the format string. That simplifies the code by using built-in functionality instead of a custom parser and is more in-line with current usage in Python 3, making it more intuitive for new users (and more in-line with the existing `SelectFiles` interface).

### Current behavior

The current implementation follows this scheme (taken [from the docs](https://nipype.readthedocs.io/en/0.12.0/interfaces/generated/nipype.interfaces.utility.html#rename)):
```python
>>> rename2 = Rename(format_string="%(subject_id)s_func_run%(run)02d")
>>> rename2.inputs.in_file = "functional.nii"
>>> rename2.inputs.keep_ext = True
>>> rename2.inputs.subject_id = "subj_201"
>>> rename2.inputs.run = 2
>>> res = rename2.run()          
>>> res.outputs.out_file         
'subj_201_func_run02.nii'     
```

The format keys are extracted with a custom regex parser [here](https://github.com/nipy/nipype/blob/b356ad16828afd4c63f862dfc72fa7e841863391/nipype/interfaces/utility/base.py#L231)
```python
class  Rename(...):
    ...
    def __init__(self, format_string=None, **inputs):
        super(Rename, self).__init__(**inputs)
        if format_string is not None:
            self.inputs.format_string = format_string
            self.fmt_fields = re.findall(r"%\((.+?)\)", format_string)
            add_traits(self.inputs, self.fmt_fields)
        else:
            self.fmt_fields = []
    ...
```

At the very end of the `_rename` the format cast would also need to be replaced:

```python
        return fmt_string % fmt_dict
```

### New behavior

With the new notation the above example would look like this:

```python
>>> rename2 = Rename(format_string="{subject_id:s}_func_run{run:02d}")
...
```

In my opinion this is easier to read and understand, because it is clearer what belongs to the keyword, what is the format specifier for the keyword and belongs to the static string. At the same time it is more intuitive for users that are used to the more modern curly bracket `{}` notation (like me :smile_cat:).

Instead of the regex parser, you could use [this suggestion](https://stackoverflow.com/questions/13037401/get-keys-from-template)

```python
# new import
from string import Formatter
...
class  Rename(...):
    ...
    def __init__(self, format_string=None, **inputs):
        super(Rename, self).__init__(**inputs)
        if format_string is not None:
            self.inputs.format_string = format_string
            # using list comprehension...
            self.fmt_fields = [i[1] for i in Formatter().parse(format_string)  if i[1] is not None]
            add_traits(self.inputs, self.fmt_fields)
        else:
            self.fmt_fields = []
    ...
```

Of course the line could be further simplified or hidden behind a function for improved readability. 

The return from the `_rename` method would look like this:

```python
        return fmt_string.format(fmt_dict)
```

### Further remarks

For backwards compatibility it would be possible to keep the regex parser around and automatically detect, which type of template is used or even use a different input attribute for the new type of formatter. That would of course make the code more complicated. You could also deprecate the older version and remove it in the future... or simply change it for nipype 2.0/pydra, breaking nipype1 code.

I am suggesting this, because after reading the docs on the `Rename` interface, I implicitly assumed it would use curly braces and the more modern `.format` method. This caused problems down the road (nipype did not find the format keys) and took me a bit of time to figure out. You are already using the `{}` format string notation in the `SelectFiles` interface and it would make sense to use the same rather than different notations/solutions for more or less identical problems. 

If you wish, I can make the PR myself. I just wanted to hear your opinions first, before implementing it.

Best,
Daniel

**Comment by dafrose** [comment]
For compatibility with the old notation, it might make sense to introduce a new attribute and have a custom parser translate input to the old attribute into the new format. That way nothing would break. 

If you do not want to change this in nipype (in contrast to pydra), you could also look for curly braces in the string and throw an error or warning, informing the user of his or her probable mistake.

**Comment by effigies** [comment]
We don't want to break old workflows. I would just make a new interface rather than overload the old one, and improve the doc strings so it's clear how to use each.

**Comment by dafrose** [comment]
I just found this template for an automatic translator from old-style `%()` string templates to new-style `{}` templates: https://gist.github.com/tonyseek/7800650

It's not perfect, but could be used as a starting point to unify the interface. 

Regarding breaking workflows @effigies : I did not find any mention over at pydra, so has this been implemented there yet? Either way, it might make sense to keep the old behaviour in nipype1 and warn about it rather than changing or translating it, whereas pydra/nipype2 could use the new style. Anything that does automatic translation from nipype1 to nipype2 should be able to handle this sort of change.

**Comment by effigies** [comment]
Yeah, I'm okay with making that the default in nipype2. I don't think it exists there yet in any case.
